runtime/src/main/java/com/dylibso/chicory/runtime/ByteArrayMemory.java

```diff
@@ -0,0 +1,350 @@
+package com.dylibso.chicory.runtime;
+
+import static com.dylibso.chicory.runtime.ConstantEvaluators.computeConstantValue;
+import static java.lang.Math.min;
+
+import com.dylibso.chicory.runtime.alloc.DefaultMemAllocStrategy;
+import com.dylibso.chicory.runtime.alloc.MemAllocStrategy;
+import com.dylibso.chicory.wasm.ChicoryException;
+import com.dylibso.chicory.wasm.UninstantiableException;
+import com.dylibso.chicory.wasm.types.ActiveDataSegment;
+import com.dylibso.chicory.wasm.types.DataSegment;
+import com.dylibso.chicory.wasm.types.MemoryLimits;
+import com.dylibso.chicory.wasm.types.PassiveDataSegment;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.nio.BufferOverflowException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.function.Function;
+
+/**
+ * Represents the linear memory in the Wasm program. Can be shared
+ * reference b/w the host and the guest.
+ */
+public final class ByteArrayMemory implements Memory {
+    // trick from:
+    // https://stackoverflow.com/a/65276765/7898052
+    private static final VarHandle SHORT_ARR_HANDLE =
```

Comment:
We have been reluctant to use `VarHandle` and `MethodHandle` in the past, in this case I believe that the usage is justified:

- there is an alternative `Memory` implementation not leveraging those
- is the fastest way to access the underlying `byte[]` (confirmed benchmarking on Photon)
- the usage is extremely limited and scoped

Reply:
We use this in Trino and it works well.

Reply:
Thanks for confirming, slightly surprised it works OOB with Android, cc. @yigit

Reply:
nice. I've never dabbed into these areas so i have no idea.
One thing though, we are running android tests only with very recent versions so behavior might be different but since you have this abstracted, it would be possible to have a different default Memory implementation for older API levels if necessary.

Reply:
@yigit `ByteBufferMemory` should work on Android too :)